bots/topological/src/main/java/org/openjdk/skara/bots/topological/TopologicalBot.java

```diff
@@ -0,0 +1,244 @@
+            if (Files.exists(depsFile)) {
+                Files.lines(depsFile)
+                        .forEach(dep -> deps.add(new Edge(new Branch(dep), branch)));
+            } else {
+                deps.add(new Edge(new Branch("master"), branch));
+            }
+        }
+        return tsort(deps).stream()
+            .filter(branch -> !branch.name().equals("master"))
+            .collect(Collectors.toList());
+    }
+
+    private void mergeIfAhead(Repository repo, Branch branch, String parent) throws IOException {
+        if (log(repo, branch.name(), parent).count() != 0) {
+            var fromHash = repo.resolve(parent).orElseThrow();
+            var isFastForward = repo.isAncestor(repo.head(), fromHash);
+            repo.merge(fromHash);
+            if (!isFastForward) {
+                log.info("Merged " + parent + " into " + branch);
+                repo.commit("Automatic merge with " + parent, "topobot", "");
+            } else {
+                log.info("Fast forwarded " + branch + " to " + parent);
+            }
+            log.info("merge with " + parent + " succeeded. The following commits will be pushed:\n"
+                    + log(repo, "origin/" + branch.name(), branch.name()).collect(Collectors.joining("\n", "\n", "\n")));
+            try {
+                repo.push(repo.head(), hostedRepo.getUrl(), branch.name());
+            } catch (IOException e) {
+                log.severe("Pusing failed! Aborting...");
+                repo.abortMerge();
+                throw e;
+            }
+        }
+    }
+
+    private static Stream<String> log(Repository repo, String targetRef, String fromRef) throws IOException {
+        var process = new ProcessBuilder()
+                .command("git", "log", targetRef + ".." + fromRef, "--")
+                .directory(repo.root().toFile())
+                .start();
+        try {
+            int exit = process.waitFor();
+            if (exit != 0) {
+                throw new IOException("Unexpected exit code: " + exit + "\n\n"
+                        + new BufferedReader(new InputStreamReader(process.getErrorStream()))
+                            .lines()
+                            .collect(Collectors.joining("\n")));
+            }
+        } catch (InterruptedException e) {
+            throw new IOException(e);
+        }
+
+        return new BufferedReader(new InputStreamReader(process.getInputStream())).lines();
+    }
+
+    private static class Edge {
+        final Branch from;
+        final Branch to;
+
+        Edge(Branch from, Branch to) {
+            this.from = from;
+            this.to = to;
+        }
+
+        @Override
+        public String toString() {
+            return "Edge{" +
+                    "from='" + from + '\'' +
+                    ", to='" + to + '\'' +
+                    '}';
+        }
+    }
+
+    private static List<Branch> tsort(List<Edge> edges) {
+        List<Edge> eCopy = new ArrayList<>(edges);
```

Comment:
It seems to me that this methods is walking all the edges and trying to find cycles, and leaves. This probably works, but can be quadratic in cost. There are better alternatives out there - some of which already tried and tested which are part of the javac compiler itself. I suggest maybe directly using this?

http://hg.openjdk.java.net/jdk/jdk/file/c16208de74da/src/jdk.compiler/share/classes/com/sun/tools/javac/util/GraphUtils.java

There, you will also find interfaces for Node and the likes. The topo sort algo used there is routinely used in javac type inference, so it's been rather heavily tested :-)

Reply:
On the other hand I don't expect a branch to have more than lets say five other branches it depends on, so I think a quadratic cost could work here :smiley:

Reply:
Yep, it's pretty quick and dirty implementation of finding any nodes without incoming edges and then adding them to the sorted list and removing from the graph. Finally, we need to once more find all leaves, since those will always have incoming edges.

It doesn't seem like perf would be a problem, and the GraphUtils impl seems a lot more extensive. So, since it works (just done writing tests), maybe stick with Quick & Dirty™ for now?

Reply:
I'm fine with quick and dirty